common/channel/src/lib.rs

```diff
@@ -77,27 +113,50 @@ impl<T> FusedStream for Receiver<T> {
 impl<T> Stream for Receiver<T> {
     type Item = T;
 
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
-        let poll = Pin::new(&mut self.value).poll_next(cx);
-        if let Poll::Ready(Some(_)) = poll {
-            self.gauge.dec();
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        loop {
+            match Pin::new(&mut self.inner).poll_next(cx) {
+                Poll::Ready(Some(msg)) => {
+                    self.gauge.dec();
+                    // If the message times out, it gets dropped
+                    if Instant::now().duration_since(msg.entry_ts) > self.timeout {
+                        continue;
+                    } else {
+                        return Poll::Ready(Some(msg.value));
+                    }
+                }
+                Poll::Ready(None) => {
+                    return Poll::Ready(None);
+                }
+                Poll::Pending => {
+                    return Poll::Pending;
+                }
+            }
         }
-        poll
     }
 }
 
 /// Similar to `mpsc::channel`, `new` creates a pair of `Sender` and `Receiver`
 pub fn new<T>(size: usize, gauge: &IntGauge) -> (Sender<T>, Receiver<T>) {
+    new_with_timeout(size, gauge, Duration::new(MAX_TIMEOUT, 0))
+}
+
+pub fn new_with_timeout<T>(
+    size: usize,
+    gauge: &IntGauge,
+    timeout: Duration,
```

Comment:
From us talking with @aching, different messages will have different timeout deadlines that change depending on how close we are to the round timeout. It seems like this is a static timeout for all messages in the queue? Instead each message could add its own `deadline: Instant` field or something? I guess we would have to change the send API then...

Reply:
Right, as the messages are opaque to network layer, consensus will have to get the messages first and then inspect them to know how long the timeout they want to set, which is like a chicken or the egg problem. After talking to @zekun000, it seems consensus team will build another layer to inspect the messages and drop unnecessary ones at their layer. So the timeout we added here is a rough bound to filter out obvious stale messages before reaching to consensus layer.